### ISSUE TITLE
NPE reading TIFF Metadata when BitsPerSample not set

### ISSUE THREAD

**Issue by myronu** [haraldk/TwelveMonkeys#254]
In our code we read the size and DPI of TIFF images using code like this:

```
imageMetadata = reader.getImageMetadata(iis); // iis is an ImageInputStream
Element root = (Element) imageMetadata.getAsTree("javax_imageio_1.0");

```
This code will fail with this NPE if the image doesn't have Bits/Sample (258) or Samples/Pixel (277) set. It does not fail using the reader from jai_imageio but instead just returns Bits/Sample = 1.

```
java.lang.NullPointerException
	at com.twelvemonkeys.imageio.plugins.tiff.TIFFImageMetadata.getStandardChromaNode(TIFFImageMetadata.java:330)
	at javax.imageio.metadata.IIOMetadata.getStandardTree(Unknown Source)
	at com.twelvemonkeys.imageio.AbstractMetadata.getAsTree(AbstractMetadata.java:75)

```
I can provide a sample image privately, but don't want to post it here. tiffinfo reports:
TIFF Directory at offset 0x8 (8)
  Subfile Type: (0 = 0x0)
  Image Width: 14400 Image Length: 11100
  Resolution: 600, 600 pixels/inch
  Compression Scheme: CCITT Group 4
  Photometric Interpretation: min-is-white
  Orientation: row 0 top, col 0 lhs
  Rows/Strip: 11100
  Planar Configuration: single image plane

**Comment by haraldk** [comment]
Hi myronu,

Thanks for reporting!

Seems you are right, there is a bug. SamplesPerPixel defaults to 1, if unset according to the spec, the metadata should behave accordingly. I've created a local fix for this, will push as asap.

I've created a fix for this, but it would be nice to have a sample image to verify. You can send it to my private email.


Best regards,

-- 
Harald K

**Comment by haraldk** [comment]
Fixed in b129117 (master), will be released as 3.3 later.

-- 
Harald K

**Comment by myronu** [comment]
Attached is the TIFF image I was working with.

Please also note that I tried opening this image using code I compiled from the latest source (prior to your changes today) using JDK 1.8.0_73 and got this RLE exception. I did not get that exception under 3.2.1 so either something didn’t compile correctly or there’s a bug in the latest code.

Caused by: java.io.IOException: Unknown code in Huffman RLE stream
                at com.twelvemonkeys.imageio.plugins.tiff.CCITTFaxDecoderStream.decodeRun(CCITTFaxDecoderStream.java:349)
                at com.twelvemonkeys.imageio.plugins.tiff.CCITTFaxDecoderStream.decode2D(CCITTFaxDecoderStream.java:182)
                at com.twelvemonkeys.imageio.plugins.tiff.CCITTFaxDecoderStream.decodeRowType6(CCITTFaxDecoderStream.java:271)
                at com.twelvemonkeys.imageio.plugins.tiff.CCITTFaxDecoderStream.decodeRow(CCITTFaxDecoderStream.java:283)
                at com.twelvemonkeys.imageio.plugins.tiff.CCITTFaxDecoderStream.fetch(CCITTFaxDecoderStream.java:114)
                at com.twelvemonkeys.imageio.plugins.tiff.CCITTFaxDecoderStream.read(CCITTFaxDecoderStream.java:435)
                at com.twelvemonkeys.io.LittleEndianDataInputStream.readFully(LittleEndianDataInputStream.java:423)
                at com.twelvemonkeys.io.LittleEndianDataInputStream.readFully(LittleEndianDataInputStream.java:396)
                at com.twelvemonkeys.imageio.plugins.tiff.TIFFImageReader.readStripTileData(TIFFImageReader.java:1386)
                at com.twelvemonkeys.imageio.plugins.tiff.TIFFImageReader.read(TIFFImageReader.java:900)

**Comment by Schmidor** [comment]
The "java.io.IOException: Unknown code in Huffman RLE stream" was caused by my recent changes to speed up CCITT reading and is fixed with #255 
This file should be another good test case for that.

**Comment by haraldk** [comment]
@myronu Sorry, attachments sent to github using email seems to be just stripped.

Can you resend to harald DOT kuhr AT Google's popular email service DOT com? :-)

-- 
Harald K

**Comment by haraldk** [comment]
@Schmidor Ahem... I think #255 is already in Master (I integrated it earlier today). Are you sure that is the fix, or is that the cause? :-)

-- 
Harald K

**Comment by Schmidor** [comment]
@haraldk I think he compiled it prior to the merge. This exception on some documents I couln't submit was the cause for writing #255.

**Comment by myronu** [comment]
I downloaded a new JAR a few minutes ago and can now open the image in our application and read the metadata with your latest changes from #254 and #255

**Comment by haraldk** [comment]
Okay, if it works, it works. 

#255 was merged before I pushed the fix for #254, so I still don't get it... But I guess all is fine then. ;-) 

.k
